scarb/src/compiler/profile.rs

```diff
@@ -1,19 +1,172 @@
+use anyhow::{anyhow, ensure, Result};
 use serde::{Deserialize, Serialize};
 use smol_str::SmolStr;
 
-use crate::core::Config;
 #[cfg(doc)]
 use crate::core::Target;
-use crate::flock::Filesystem;
+use crate::core::{ManifestCompilerConfig, TomlCairo};
 
-/// Profile settings used to determine which compiler flags to use for a [`Target`].
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct Profile {
-    pub name: SmolStr,
-}
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq, Hash)]
+pub struct Profile(SmolStr);
 
 impl Profile {
-    pub fn target_dir<'c>(&self, config: &'c Config) -> Filesystem<'c> {
-        config.target_dir().child(self.name.as_str())
+    pub const RELEASE: Self = Self(SmolStr::new_inline("release"));
+    pub const DEV: Self = Self(SmolStr::new_inline("dev"));
+
+    pub fn new(name: SmolStr) -> Result<Self> {
+        // Some validations are required to ensure profile name can be used as a valid subdirectory.
+        ensure!(
+            name.as_str() != "",
+            "cannot use empty string as profile name"
+        );
+        ensure!(
+            !vec!["_", "package", "build", "debug", "doc", "test"].contains(&name.as_str()),
+            format!("profile name `{name}` is not allowed")
+        );
+        ensure!(
+            !name.to_string().starts_with(".."),
+            format!("profile name cannot start with `..` prefix")
+        );
+        ensure!(
+            name.chars().all(|c| c.is_ascii_alphanumeric() || c == '-'),
+            format!("profile name `{name}` is not allowed, only alphanumeric characters and `-` can be used")
+        );
+        Ok(Self(name))
+    }
+
+    pub fn name(&self) -> SmolStr {
+        self.0.clone()
+    }
+
+    pub fn is_dev(&self) -> bool {
+        self.0.as_str() == "dev"
+    }
+    pub fn is_release(&self) -> bool {
+        self.0.as_str() == "release"
+    }
+    pub fn is_custom(&self) -> bool {
+        !self.is_dev() && !self.is_release()
+    }
+}
+
+impl Default for Profile {
+    fn default() -> Self {
+        Self::DEV
+    }
+}
+
+impl From<SmolStr> for Profile {
+    fn from(name: SmolStr) -> Self {
+        match name.as_str() {
+            "dev" => Self::DEV,
+            "release" => Self::RELEASE,
+            _ => Self::new(name).unwrap(),
+        }
+    }
+}
+
+impl From<Profile> for SmolStr {
+    fn from(profile: Profile) -> Self {
+        match profile.0.as_ref() {
+            "dev" => "dev".into(),
+            "release" => "release".into(),
+            _ => profile.name(),
+        }
+    }
+}
+
+pub trait DefaultForProfile {
+    fn default_for_profile(profile: &Profile) -> Self;
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ProfileBuilder {
```

Comment:
I don't like this type. It sounds like a bad, leaky abstraction:
1. Leaky, because you import unrelated types into this module and type.
2. Also it looks more Javaesque than Rusty. This is object-oriented rather than data-.
 
I think what would play better here would be a mechanism for _merging_ two objects, which can be done on any `T: Serialize + Deserialize`: you serialize object A (here: default for profile) to TOML Value. Then you merge user globals into it. And finally you merge profile overrides.